zcash_client_backend/src/data_api.rs

```diff
@@ -1583,11 +1583,20 @@ impl AccountBirthday {
 /// - If [`WalletWrite::import_account_hd`] is used to import accounts with non-sequential
 ///   ZIP 32 account indices from the same seed, a call to [`WalletWrite::create_account`]
 ///   will use the ZIP 32 account index just after the highest-numbered existing account.
-/// - If an account is imported via [`WalletWrite::import_account_ufvk`], and then a later
-///   call to either [`WalletWrite::create_account`] or [`WalletWrite::import_account_hd`]
-///   would produce the same UFVK, an error will be returned.
-///   - A future change to this trait might introduce a method to "upgrade" an imported
-///     account with derivation information. See [zcash/librustzcash#1284] for details.
+/// - If an account is added to the wallet, and then a later call to one of the methods
+///   would produce the same UFVK, an error will be returned. While any combination of
+///   method calls can cause this, there are two main ways it can occur that users of the
+///   `WalletWrite` trait should watch out for:
+///   - An account is created via [`WalletWrite::create_account`] with an auto-selected
+///     ZIP 32 account index, and that index is later imported explicitly via either
+///     [`WalletWrite::import_account_hd`] or [`WalletWrite::import_account_ufvk`].
+///   - An account is imported via [`WalletWrite::import_account_ufvk`], and the ZIP 32
+///     account index corresponding to that account's UFVK is later imported either
+///     implicitly via [`WalletWrite::create_account`] or explicitly via
+///     [`WalletWrite::import_account_hd`].
```

Comment:
I don't understand why these two cases are being called out. Why is it less likely that you call the same method twice with the same arguments, for instance?

Reply:
Note that this is independent of the issue described in https://github.com/zcash/librustzcash/issues/1474#issuecomment-2261507844 that we don't currently detect collisions between non-identical UFVKs with shared components. The three cases where an account creation method collides with another invocation of itself, and the two remaining cases not covered above (`import_account_hd`, then `create_account` or `import_account_ufvk`) can all currently occur.

Reply:
> I don't understand why these two cases (which represent 4 out of the 9 possible combinations of (method to create first account, method to create colliding account)) are being specifically called out.

Because those are (I think) the most likely cases that users of the `WalletWrite` trait may accidentally encounter without realising.

> Why is it less likely that you call the same method twice with the same arguments, for instance?

Because:
- this cannot happen with the `create_account` API.
- this is very obvious for the `import_account_*` APIs.

Reply:
> Note that this is independent of the issue described in https://github.com/zcash/librustzcash/issues/1474#issuecomment-2261507844 that we don't currently detect collisions between non-identical UFVKs with shared components.

Yes I'm well aware of that, and this PR is not attempting to address that issue.

Reply:
> > Why is it less likely that you call the same method twice with the same arguments, for instance?
> 
> Because:
> 
> * this cannot happen with the `create_account` API.

That is not correct. Calling `create_account` with the same seed will result in a duplicate UFVK. You may say "the intended use of the API is to call it with a fresh random seed" (in which case a collision is obviously vanishingly unlikely), but nothing requires that and the API docs don't mention it.

> * this is very obvious for the `import_account_*` APIs, and thus I don't think needs to be called out here (for the complexity reason I give above).

The wording that avoids calling these 4 out of 9 cases as in any way special is *less* complex. As it is currently written, it immediately makes me wonder "what is different about these cases, that I am supposed to pay special attention to them?" And the answer is, nothing.

Reply:
> > > Why is it less likely that you call the same method twice with the same arguments, for instance?
> > 
> > Because:
> > 
> > * this cannot happen with the `create_account` API.
>
> That is not correct. Calling `create_account` with the same seed will result in a duplicate UFVK. You may say "the intended use of the API is to call it with a fresh random seed" (in which case a collision is obviously vanishingly unlikely), but nothing requires that and the API docs don't mention it.

That is not correct. Calling `create_account` with the same seed will result in the next UFVK being generated, corresponding to the next unused ZIP 32 account index (or an error if you've hit the end of the account range). Sequential calls to `create_account` with the same arguments _never_ result in the same changes to the wallet database.

Reply:
The doc still omits cases. For example, you can import an account with `import_account_ufvk` first, and then `create_account` with the same seed can fail because `max_zip32_account_index` will not find the UFVK-imported account, since the wallet doesn't yet know the seed.

Reply:
That's... literally the case documented in the bullet point that this thread is anchored to?
> - **An account is imported via `WalletWrite::import_account_ufvk`, and the ZIP 32 account index corresponding to that account's UFVK is later imported either implicitly via `WalletWrite::create_account`** or explicitly via `WalletWrite::import_account_hd`.